fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/LocalFileBinary.java

```diff
@@ -0,0 +1,150 @@
+
+        try {
+            /* that this is a proxy or redirect has already been set before we get here
+             */
+            final Node descNode = getDescriptionNode();
+            final Node contentNode = getNode();
+
+            LOGGER.debug("HERE HERE HERE HERE HERE");
+            if (contentNode.canAddMixin(FEDORA_BINARY)) {
+                contentNode.addMixin(FEDORA_BINARY);
+            }
+
+            if (originalFileName != null) {
+                descNode.setProperty(FILENAME, originalFileName);
+            }
+
+            if (contentType == null) {
+                descNode.setProperty(HAS_MIME_TYPE, DEFAULT_MIME_TYPE);
+            } else {
+                descNode.setProperty(HAS_MIME_TYPE, contentType);
+            }
+
+            // Store the required jcr:data property
+            contentNode.setProperty(JCR_DATA, "");
+
+            LOGGER.debug("Created content node at path: {}", contentNode.getPath());
+
+            // Ensure provided checksums are valid
+            final Collection<URI> nonNullChecksums = (null == checksums) ? new HashSet<>() : checksums;
+            verifyChecksums(nonNullChecksums);
+
+            decorateContentNode(contentNode, descNode, nonNullChecksums, getContentSize());
+            FedoraTypesUtils.touch(contentNode);
+            FedoraTypesUtils.touch(descNode);
+
+            LOGGER.debug("Set local file content at path: {}", getResourceLocation());
+
+        } catch (final RepositoryException e) {
+            throw new RepositoryRuntimeException(e);
+        }
+    }
+
+    private static void decorateContentNode(final Node dsNode, final Node descNode, final Collection<URI> checksums,
+        final long size)
+        throws RepositoryException {
+        if (dsNode == null) {
+            LOGGER.warn("{} node appears to be null!", JCR_CONTENT);
+            return;
+        }
+
+        if (dsNode.hasProperty(JCR_DATA)) {
+            // Store checksums on node
+            final String[] checksumArray = new String[checksums.size()];
+            checksums.stream().map(Object::toString).collect(Collectors.toSet()).toArray(checksumArray);
+            descNode.setProperty(CONTENT_DIGEST, checksumArray);
+
+            // Store the size of the file
+            contentSizeHistogram.update(size);
+            descNode.setProperty(CONTENT_SIZE, size);
+
+            LOGGER.debug("Decorated local file data property at path: {}", dsNode.getPath());
+        }
+    }
+
+    @Override
+    public String getMimeType() {
+        return getMimeTypeValue();
+    }
+
+    /*
+     * (non-Javadoc)
+     * @see org.fcrepo.kernel.modeshape.FedoraBinaryImpl#getContentSize()
+     */
+    @Override
+    public long getContentSize() {
+        final long sizeValue = super.getContentSize();
+        if (sizeValue > -1L) {
+            return sizeValue;
+        }
+        final File file = new File(getResourceUri().getPath());
+        setContentSize(file.length());
```

Comment:
Pretty sure I wrote this, but this is a bad idea, we don't want to be changing a property of the description in a getter. A more appropriate place to retrieve the file size and set it would be in setExternalContent (if its not a memento). Maybe override setExternalContent, have it call super.setExternalContent, then call setContentSize.

Reply:
I tried to change this and it needs more work.  Or rather, I tried to take this line out and do this instead: `setContentSize(getContentSize());` inside of setExternalContent.  It resulted in tests failing in a weird way. I can make a ticket to come back to this because I really want to get this PR in and then change things from there. Sound okay?